src/svxlink/svxlink/contrib/SipLogic/SipLogic.conf.5
.TH SIPLOGIC.CONF 5 "November 2022" Linux "File Formats"
.
.SH NAME
.
SipLogic.conf \- SipLogic configuration file for the SvxLink server
.
.SH DESCRIPTION
.
.B svxlink
is a general purpose voice service system for ham radio use. This man-page
describe the SvxLink server configuration file format for the SipLogic logic
core.
.
.SH CONFIGURATION VARIABLES
.
Here is the description of all configuration variables that the SipLogic
SvxLink logic core understands. The configuration variables are described
section for section.
.
.SS SipLogic
.
The SipLogic is used to connect to a Sip-based communication server like
Astrisk.
.TP
.B TYPE
The type for a sip logic core is always
.BR Sip .
.TP
.B SIPSERVER
The ip address or the hostname of the Sip communication server.
.TP
.B SIPPORT
The udp port that is used for communication (normally 5060).
.TP
.B CALLERNAME
The name of the station that is displayed instead of the number.
.TP
.B USERNAME
The username that is used to connect to the Sip communication server.
.TP
.B PASSWORD
The password for the USERNAME to connect to the Sip communication server.
.TP
.B SIPREGISTRAR
The sip registrar or better kown as proxy (e.g. "fritz.box" for AVM Fritzboxes)
.TP
.B SIPSCHEMA
The authentication scheme (e.g. "digest")
.TP
.B SIPEXTENSION
The extension to used when connecting to a Sip communication server.
.TP
.B AUTOANSWER
If set the call will be pickuped immediately when received.
AUTOANSWER=(0|1)
.TP
.B AUTOCONNECT
Define a party that will being connected after upon startup of SvxLink.
AUTOCONNECT=user123@sipaccount.com
.TP
.B REJECT_INCOMING
Reject all incoming phone calls where the remote number match this regular
expression. For more information on the syntax, seek for
.B REGULAR EXPRESSIONS
.TP
.B ACCEPT_INCOMING
Only accept incoming phone calls where the remote number match this regular
expression.
For more information on the syntax, seek for
.B REGULAR EXPRESSIONS
.TP
.B REJECT_OUTGOING
Reject all outgoing phone calls where the remote number match this regular
expression. For more information on the syntax, seek for
.B REGULAR EXPRESSIONS
.TP
.B ACCEPT_OUTGOING
Only accept outgoing phone calls where the remote number match this regular
expression.
For more information on the syntax, seek for
.B REGULAR EXPRESSIONS
.TP
.B SEMI_DUPLEX (ToDo)
If enabled no VOX squelch is used for Sip clients. That means that the audio
stream from the SipLogic is sent continuously to the connected SvxLink nodes.
That is only meaningful if all connected nodes are repeaters.
SEMI_DUPLEX=(0|1)
.TP
.B VOX_FILTER_DEPTH
see "Local Receiver Section"
.TP
.B VOX_THRESH
see "Local Receiver Section"
.TP
.B SQL_START_DELAY
see "Local Receiver Section"
.TP
.B SQL_DELAY
see "Local Receiver Section"
.TP
.B SQL_HANGTIME
see "Local Receiver Section"
.TP
.B SIP_LOGLEVEL
The debug output of the pjsip library (0..6)
.TP
.B JITTER_BUFFER_DELAY
A jitter buffer is used to prevent gaps in the audio when the network
connection do not provide a steady flow of data. Set this configuration
variable to the number of milliseconds to buffer before starting to process the
audio. Default: 0.
.TP
.B REG_TIMEOUT
Time in seconds to renewal the registration at the sip server.
.TP
.B CALL_TIMEOUT
Time in seconds after an outgoing call will be terminate when the party will not
pickup the call.
.TP
.B SIP_CTRL_PTY
Using this configuration variable it is possible to specify a path to a UNIX 98
PTY that allows a dtmf control of each single SvxLink logic. SvxLink will create
a softlink to the actual slave PTY. For that reason, SvxLink must have write
permissions in the directory where the softlink should be created. Sending
commands to the PTY is as simple as doing a
.B "echo 'C1#' > /path/to/pty"
after starting SvxLink. The device works bidirectional, received dtmf characters
(from local application) are output via this interface.
Example: SIP_CTRL_PTY=/dev/shm/dtmf_ctrl
.TP
.B SIP_PREAMP
The incoming signal from the Sip network will be amplified by the specified
number of dB. This can be used as a last measure if the input audio level
can't be set high enough on the analogue side. A value of 6dB will double
the signal level. Note that this is a digital amplification. Hence it will
reduce the dynamic range of the signal so usage should be avoided if possible.
It's always better to correct the audio level before sampling it.
.TP
.B SIP_LIMITER_THRESH
Set the threshold, in dBFS, for the audio limiter. The audio limiter really is
a compressor with a very steep compression ratio like 10:1. The limiter is
used to help keeping down the audio level from Sip network for overmodulating
stations. Try values from -6 to -12 if the incoming Sip audio is to high. Set
to 0 to deactivate it.
.TP
.B PHONENUMBER_TO_TG
Comma-separated list of assignments of phonenumber(full||substring):Talkgroup.
When a phone call is received, the call is assigned to the defined TG according
to the recognized number. This can be used to prevent the transfer of an SIP
call to a network or to carry out a country-specific assignment. The number
format transmitted must match the configuration (+49 != 0049). Whole SIP
numbers as well as parts of phone numbers can be configured, with the
evaluation taking place on the left-hand side.
Example: PHONENUMBER_TO_TG=0049:262,0034:214,0043:232,0049123454543:991
.TP
.B USE_TCP
If set to 1, TCP is used as transport layer instead of UDP (default).
.TP
.B IGNORE_REGISTRATION
If set to 1, the SipLogic doesn't check if the incoming caller is registered
at the own Sip server (default: false).
.
.SH FILES
.
.TP
.IR /etc/svxlink/svxlink.d/SipLogic.conf
The SipLogic specific configuration file.
.TP
.IR /etc/svxlink/svxlink.conf
The system wide configuration file.
.TP
.IR ~/.svxlink/svxlink.conf
Per user configuration file.
.TP
.I /etc/svxlink/svxlink.d/*
Additional configuration files. Typically one configuration file per module.
.
.SH AUTHOR
.
Adi Bier (DL1HRC)
.
.SH "SEE ALSO"
.
.BR svxlink.conf (5)


src/svxlink/svxlink/SipLogic.tcl
###############################################################################
#
# SipLogic event handlers
#
###############################################################################

#
# This is the namespace in which all functions below will exist. The name
# must match the corresponding section "[SipLogic]" in the configuration
# file. The name may be changed but it must be changed in both places.
#
namespace eval SipLogic {

#
# Checking to see if this is the correct logic core
#
if {$logic_name != [namespace tail [namespace current]]} {
  return;
}


#
# Executed when the SvxLink software is started
#
proc startup {} {
  global logic_name;
  puts "starting $logic_name";
}


#
# Executed when an unknown DTMF command is entered
#
proc unknown_command {cmd} {
  playMsg "Core" "unknown_command";
  playNumber $cmd;
}


#
# Executed when an outgoing call is not permitted (dropped)
#
proc drop_outgoing_call {caller} {
  puts "Dropping outgoing call to $caller due to configuration.";
  playMsg "SipLogic" "call_not_allowed";
  playSilence 200;
}


#
# Executed when an incoming call is rejected
#
proc reject_incoming_call {caller} {
  puts "Rejecting incoming call from $caller due to configuration.";
  playMsg "SipLogic" "reject_incoming_call";
  playSilence 200;
}


#
# Executed when an entered DTMF command failed
#
proc command_failed {cmd} {
  playMsg "Core" "command_failed";
  playNumber "Core" $cmd;
}


#
# Executed when a link to another logic core is activated.
#   name  - The name of the link
#
proc activating_link {name} {
  playMsg "Core" "activating_link";
  playMsg "Core" $name;
}


#
# Executed when a link to another logic core is deactivated.
#   name  - The name of the link
#
proc deactivating_link {name} {
  playMsg "Core" "deactivating_link";
  playMsg "Core" $name;
}


#
# Executed when trying to deactivate a link to another logic core but the
# link is not currently active.
#   name  - The name of the link
#
proc link_not_active {name} {
#  Logic::link_not_active $name;
}


#
# Executed when trying to activate a link to another logic core but the
# link is already active.
#   name  - The name of the link
#
proc link_already_active {name} {
#  Logic::link_already_active $name;
}


#
# Executed once every whole minute
#
proc every_minute {} {
  Logic::every_minute;
}


#
# Executed once every second
#
proc every_second {} {
  Logic::every_second;
}


#
# Executed once every whole minute to check if it's time to identify
#
proc checkPeriodicIdentify {} {
# Logic::checkPeriodicIdentify;
}


#
# Executed if an incoming call is pickuped
#
proc pickup_call {caller} {
  puts "pickup call from: $caller";
  playSilence 500;
}


#
# Executed if an outgoing call is pending
#
proc calling {caller} {
  playSilence 500;
  playMsg "SipLogic" "outgoing_phonecall";
  playSilence 100;
  spellNumber [getCallerNumber $caller];
}


#
# Executed if an outgoing call is pending
#
proc outgoing_call {caller} {
  puts "outgoing_call $caller";
}


#
# Executed if somebody is ringing
#
proc ringing {caller} {
  puts "$caller ringing";
  playMsg "SipLogic" "ringtone";
}


#
# Executed if an incoming call is established
#
proc incoming_call {caller} {
  playSilence 500;
  playMsg "SipLogic" "incoming_phonecall";
  playSilence 100;
  playNumber [getCallerNumber $caller];
  playSilence 200;
}


#
# Executed when a DTMF digit has been received
#   digit - The detected DTMF digit
#   code  - The duration, in milliseconds, of the digit
#
# Return 1 to hide the digit from further processing in SvxLink or
# return 0 to make SvxLink continue processing as normal.
#
proc dtmf_digit_received {digit caller} {
  puts "Dtmf digit received $digit from $caller";
}


#
# Executed when a DTMF command has been received
#   cmd - The command
#
# Return 1 to hide the command from further processing is SvxLink or
# return 0 to make SvxLink continue processing as normal.
#
proc dtmf_cmd_received {cmd} {
  return 0;
}


#
# Executed when the node is being brought online after being offline
#
proc logic_online {online} {
  puts "$online";
}


#
# Executed when the party isn't at home
#
proc call_timeout {} {
  puts "Called party is not at home";
  playMsg "SipLogic" "person_not_available";
  playSilence 200;
}


#
# Executed when a call has hangup
#
proc hangup_call {uri duration} {
  puts "Hangup call $uri ($duration seconds)";
  playMsg "SipLogic" "call_terminated";
  playSilence 200;
}


#
# Registration state (code)
#
proc registration_state {server state code} {
  variable status;
  if {$state} {
    set status "registered";
  } else {
    set status "unregistered";
  }
  puts "Registration on $server, code $code ($status)";
}


#
# Returns the phone number of the remote station
# URI: <sip:12345678@sipserver.com:5060>
#        -> 12345678
#
proc getCallerNumber {uri} {
  set r [split $uri @];
  set number [split [lindex $r 0] :];
  set nr [lindex $number 1];
  if {$nr ne ""} {
    return $nr;
  } else {
    return $uri;
  }
}


#
# ToDo: Sends a reject-message to the caller via Sip
#
proc remote_reject_call {} {
  playMsg "call_rejected";
  playSilence 2500;
}


#
# ToDo: Sends a greeting message to the caller via Sip
#
proc remote_greeting {} {
  playMsg "SipLogic" "welcome";
  playSilence 200;
}


#
# Executed when an incoming call has ben picked 
#
proc incoming_call_answered {caller} {
}


#
# Executed when a call(s) has been hangup by a rf user
# by dtmf
#
proc call_hangup_by_user {} {
}


#
#
#
proc call_state_confirmed {caller} {
}


#
# Executed at unknown callstate 
#
proc unknown_callstate {caller} {
}


#
# Executed if a PJSIP_INV_STATE_EARLY event occurred
#
proc pjsip_state_early {caller} {
}


#
# Executed if a PJSIP_INV_STATE_NULL event occurred
#
proc pjsip_state_null {caller} {
}


#
# Executed on incoming sip message from a connected party
#
proc text_message_received {uri message} {
  puts "Message from $uri received:";
  puts "$message";
}


# Executed on imcoming sip message from a not connected party over
# the registered account
#
proc account_text_message_received {uri message} {
  puts "Message from $uri received:";
  puts "$message";
}


#
# Executed if a call is registered
#
#
proc call_registered {caller} {
}

# end of namespace
}


#
# This file has not been truncated
#
